dss/util/security.py

```diff
@@ -0,0 +1,128 @@
+    verified_token = verify_jwt(token)
+    if Config.get_token_info_url() in verified_token['aud']:
+        token_request = session.get(Config.get_token_info_url(),
+                                    headers={'Authorization': 'Bearer {}'.format(token)},
+                                    timeout=5)
+        verified_token.update(token_request.json())
+        if verified_token.get('email_verified') is not True:
+            logger.info("Email not verified: {verified_token}")
+            raise DSSForbiddenException()
+    else:
+        # then token is a google service credential from a trusted google project.
+        verified_token['email'] = verified_token['sub']
+        if not (verified_token.get('scope') and verified_token.get('scopes')):
+            verified_token['scopes'] = []
+    return verified_token
+
+
+def verify_jwt(token: str) -> typing.Optional[typing.Mapping]:
+    try:
+        unverified_token = jwt.decode(token, verify=False)
+        issuer = unverified_token["iss"]
+        is_authorized_issuer(issuer)
+        public_keys = get_public_keys(issuer)
+        token_header = jwt.get_unverified_header(token)
+        if Config.deployment_stage() != 'prod':
+            logging.info(f"Token expiration is not checked in non production stages. Token: {token}")
+            tok = jwt.decode(token,
+                             key=public_keys[token_header["kid"]],
+                             issuer=issuer,
+                             audience=Config.get_audience(),
+                             algorithms=allowed_algorithms,
+                             options={'verify_exp': False})
+        else:
+            tok = jwt.decode(token,
+                             key=public_keys[token_header["kid"]],
+                             issuer=issuer,
+                             audience=Config.get_audience(),
+                             algorithms=allowed_algorithms,
+                             )
+    except jwt.PyJWTError as ex:  # type: ignore
+        logger.info(f"JWT failed to validate.", exc_info=True)
+        raise DSSException(401, 'Unauthorized', 'Authorization token is invalid') from ex
+    return tok
+
+
+def is_authorized_issuer(issuer: str) -> None:
+    """
+    Must be either "https://humancellatlas.auth0.com/" or service credential from a trusted google project.
+    :param issuer: str
+    """
+    if issuer != Config.get_openid_provider() and not any(issuer.endswith(f"@{p}")
+                                                          for p in Config.get_trusted_google_projects()):
+        logger.info(f"Token issuer not authorized: {issuer}")
+        raise DSSForbiddenException()
+
+
+def is_authorized_group(group: typing.List[str], token_info: dict) -> None:
+    if (token_info.get("https://auth.data.humancellatlas.org/group") in group and
+            token_info.get('email_verified', False) is True):
+        return
+    if token_info.get("sub", '').endswith(gserviceaccount):
```

Comment:
This should not be special cased. Where possible, services should be treated the same as users for all auth intents and purposes. For now, since services can mint their own JWTs, they should be responsible for injecting their own groups to get the privileges they want. In the future we may manage that for them.

Reply:
fixed